File: Capítulo_14/Programação dinâmica/Maior subsequência comum/logestCommonSubsequencePrint.js
// Algoritmo do lcs (Longest Common Subsequence, ou Maior Subsequência Comum) com resultado
function lcs(wordX, wordY) {
    const m = wordX.length;
    const n = wordY.length;
    const l = [];
    const solution = [];
    for (let i = 0; i <= m; i++) {
        l[i] = [];
        solution[i] = [];
        for (let j = 0; j <= n; j++) {
            l[i][j] = 0;
            solution[i][j] = "0";
        }
    }
    for (let i = 0; i <= m; i++) {
        for (let j = 0; j <= n; j++) {
            if (i === 0 || j === 0) {
                l[i][j] = 0
            } else if (wordX[i - 1] === wordY[j - 1]) {
                l[i][j] = l[i - 1][j] + 1;
                solution[i][j] = "diagonal";
            } else {
                const a = l[i - 1][j];
                const b = l[i][j - 1];
                l[i][j] = a > b ? a : b; // max(a, b)
                solution[i][j] = l[i][j] === l[i - 1][j] ? "top" : "left";
            }
        }
    }
    return [l[m][n], printSolution(solution, wordX, m, n)]
}

// Função printSolution:
function printSolution(solution, wordX, m, n) {
    let a = m;
    let b = n;
    let x = solution[a][b];
    let answer = "";
    while (x !== "0") {
        if (solution[a][b] === "diagonal") {
            answer = wordX[a - 1] + answer;
            a--;
            b--;
        } else if (solution[a][b] === "left") {
            b--;
        } else if (solution[a][b] === "top") {
            a--;
        }
        x = solution[a][b];
    }
    // console.log("lcs: " + answer);
    return answer
}

// Testando algoritmo lcs
let wordX = "acbaed";
let wordY = "abcadf";
let result = lcs(wordX, wordY);

console.log(`A maior subsequência comum entre "${wordX}" e "${wordY}" tem comprimento: ${result[0]} e lcs: ${result[1]}`);

File: Capítulo_14/Programação dinâmica/Maior subsequência comum/Códigos em Python/logestCommonSubsequencePrint.py
# Algoritmo do lcs (Longest Common Subsequence, ou Maior Subsequência Comum)
def lcs(word_x, word_y):
    m = len(word_x)
    n = len(word_y)
    l = [[0 for _ in range(m + 1)] for _ in range(n + 1)] # Inicialização da matrix l
    solution = [["0" for _ in range(m + 1)] for _ in range (n + 1)] # Inicialização da matrix solution
    for i in range(m + 1):
        for j in range(n + 1):
            if i == 0 or j == 0:
                l[i][j] = 0
            elif word_x[i - 1] == word_y[j - 1]:
                l[i][j] = l[i - 1][j - 1] + 1
                solution[i][j] = "diagonal"
            else:
                a = l[i - 1][j]
                b = l[i][j - 1]
                l[i][j] = a if a > b else b # max(a, b)
                solution[i][j] = "top" if l[i][j] == l[i - 1][j] else "left"

    return l[m][n], print_solution(solution, word_x, m, n)

# Função print_solution:
def print_solution(solution, word_x, m, n):
    a = m
    b = n
    x = solution[a][b]
    answer = ""

    while x != "0":
        if solution[a][b] == "diagonal":
            answer = word_x[a - 1] + answer
            a -= 1
            b -= 1
        elif solution[a][b] == "left":
            b -= 1
        elif solution[a][b] == "top":
            a -= 1
        x = solution[a][b]
    
    # print(f"lcs: {answer}")
    return answer

# Testando algoritmo lcs
word_x = "acbaed"
word_y = "abcadf"

result = lcs(word_x, word_y)
print(f"A maior subsequência comum entre \"{word_x}\" e \"{word_y}\" tem comprimento: {result[0]} e lcs: {result[1]}") 
